aten/src/ATen/native/quantized/cpu/q_avgpool.cpp

```diff
@@ -0,0 +1,222 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/Parallel.h>
+#include <ATen/core/op_registration/op_registration.h>
+
+#include <algorithm>
+#include <limits>
+#include <vector>
+
+namespace at { namespace native {
+namespace {
+
+inline int start_index(int a, int b, int c) {
+  return (int)std::floor((float)(a * c) / b);
+}
+
+inline int end_index(int a, int b, int c) {
+  return (int)std::ceil((float)((a + 1) * c) / b);
+}
+
+template <typename underlying_t>
+inline underlying_t clamp_and_cast(int32_t value) {
+  const auto lower = std::numeric_limits<underlying_t>::lowest();
+  const auto upper = std::numeric_limits<underlying_t>::max();
+  return static_cast<underlying_t>(
+      std::max<int32_t>(lower, std::min<int32_t>(upper, value)));
+}
+
+template <typename scalar_t, typename underlying_t>
+static void adaptive_avg_pool2d_single_out_frame(scalar_t *input_p,
+                                                 scalar_t *output_p,
+                                                 int64_t sizeD,
+                                                 int64_t isizeH,
+                                                 int64_t isizeW,
+                                                 int64_t osizeH,
+                                                 int64_t osizeW,
+                                                 int64_t istrideD,
+                                                 int64_t istrideH,
+                                                 int64_t istrideW) {
+  at::parallel_for(0, sizeD, 0, [&](int64_t start, int64_t end) {
+    for (auto d = start; d < end; d++) {
+      /* loop over output */
+      int64_t oh, ow;
+      for(oh = 0; oh < osizeH; oh++) {
+        int istartH = start_index(oh, osizeH, isizeH);
+        int iendH   = end_index(oh, osizeH, isizeH);
+        int kH = iendH - istartH;
+
+        for(ow = 0; ow < osizeW; ow++) {
+          int istartW = start_index(ow, osizeW, isizeW);
+          int iendW   = end_index(ow, osizeW, isizeW);
+          int kW = iendW - istartW;
+
+          /* local pointers */
+          scalar_t *ip = input_p   + d*istrideD + istartH*istrideH + istartW*istrideW;
+          scalar_t *op = output_p  + d*osizeH*osizeW + oh*osizeW + ow;
+
+          /* compute local average: */
+          int32_t sum = 0;
+          int ih, iw;
+          for(ih = 0; ih < kH; ih++) {
+            for(iw = 0; iw < kW; iw++) {
+              int32_t val = (ip + ih*istrideH + iw*istrideW)->val_;
+              sum += val;
+            }
+          }
+
+          /* set output to local average */
+          *op = scalar_t(clamp_and_cast<underlying_t>(sum / kW / kH));
```

Comment:
Where is the output quantized back to 8 bits?

Reply:
Why do we need to quantize it back?